salt/channel/client.py

```diff
@@ -0,0 +1,555 @@
+                load, tries=tries, timeout=timeout, raw=raw
+            )
+        raise salt.ext.tornado.gen.Return(ret)
+
+    def close(self):
+        """
+        Since the message_client creates sockets and assigns them to the IOLoop we have to
+        specifically destroy them, since we aren't the only ones with references to the FDs
+        """
+        if self._closing:
+            return
+        log.debug("Closing %s instance", self.__class__.__name__)
+        self._closing = True
+        self.transport.close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self.close()
+
+
+class AsyncPubChannel:
+    """
+    Factory class to create subscription channels to the master's Publisher
+    """
+
+    async_methods = [
+        "connect",
+        "_decode_messages",
+    ]
+    close_methods = [
+        "close",
+    ]
+
+    @classmethod
+    def factory(cls, opts, **kwargs):
+        import salt.ext.tornado.ioloop
+        import salt.crypt
+
+        # Default to ZeroMQ for now
+        ttype = "zeromq"
+
+        # determine the ttype
+        if "transport" in opts:
+            ttype = opts["transport"]
+        elif "transport" in opts.get("pillar", {}).get("master", {}):
+            ttype = opts["pillar"]["master"]["transport"]
+
+        if "master_uri" not in opts and "master_uri" in kwargs:
+            opts["master_uri"] = kwargs["master_uri"]
+
+        # switch on available ttypes
+        if ttype == "detect":
+            opts["detect_mode"] = True
+            log.info("Transport is set to detect; using %s", ttype)
+
+        io_loop = kwargs.get("io_loop")
+        if io_loop is None:
+            io_loop = salt.ext.tornado.ioloop.IOLoop.current()
+
+        auth = salt.crypt.AsyncAuth(opts, io_loop=io_loop)
+        transport = salt.transport.publish_client(opts, io_loop)
+        return cls(opts, transport, auth, io_loop)
+
+    def __init__(self, opts, transport, auth, io_loop=None):
+        self.opts = opts
+        self.io_loop = io_loop
+        self.auth = auth
+        self.tok = self.auth.gen_token(b"salt")
+        self.transport = transport
+        self._closing = False
+        self._reconnected = False
+        self.event = salt.utils.event.get_event("minion", opts=self.opts, listen=False)
+
+    @property
+    def crypt(self):
+        if self.auth:
+            return "aes"
+        return "clear"
+
+    @salt.ext.tornado.gen.coroutine
+    def connect(self):
+        """
+        Return a future which completes when connected to the remote publisher
+        """
+        try:
+            import traceback
+
+            if not self.auth.authenticated:
+                yield self.auth.authenticate()
+            # if this is changed from the default, we assume it was intentional
+            if int(self.opts.get("publish_port", 4506)) != 4506:
+                publish_port = self.opts.get("publish_port")
```

Comment:
**question** why do we `int` for the comparison but not for setting?

Reply:
`publish_port` gets used further down the stack in a couple different ways. The zeromq transport uses it to format a url string. The tcp transport casts it to an int when it's needed. I don't expect the rabbitmq transport will even need  `publish_port`. The connection details needed very quite a bit from transport to transport. We've discussed the need to abstract these connection details a bit more but I didn't want to blow up the scope of this re-factor too much. I think we'll want to take make another pass that focuses on cleaning up the config for transports. I think channels should be passing in some generic connection details class to transports.

Reply:
I've captured this in #61362